std/experimental/allocator/building_blocks/bitmapped_block.d

```diff
@@ -0,0 +1,1437 @@
+        if (blocks >= blocksAtATime)
+        {
+            foreach (i; 0 .. blocks / blocksAtATime - 1)
+            {
+                auto b = a.allocate(bs * blocksAtATime);
+                assert(b.length == bs * blocksAtATime, text(i, ": ", b.length));
+                (cast(ubyte[]) b)[] = 0xff;
+                a.expand(b, blocksAtATime * bs)
+                    || assert(0, text(i));
+                (cast(ubyte[]) b)[] = 0xfe;
+                assert(b.length == bs * blocksAtATime * 2, text(i, ": ", b.length));
+                a.reallocate(b, blocksAtATime * bs) || assert(0);
+                assert(b.length == bs * blocksAtATime, text(i, ": ", b.length));
+            }
+        }
+    }
+
+    testAllocateAll!(1)(0, 1);
+    testAllocateAll!(1)(8, 1);
+    testAllocateAll!(4096)(128, 1);
+
+    testAllocateAll!(1)(0, 2);
+    testAllocateAll!(1)(128, 2);
+    testAllocateAll!(4096)(128, 2);
+
+    testAllocateAll!(1)(0, 4);
+    testAllocateAll!(1)(128, 4);
+    testAllocateAll!(4096)(128, 4);
+
+    testAllocateAll!(1)(0, 3);
+    testAllocateAll!(1)(24, 3);
+    testAllocateAll!(3008)(100, 1);
+    testAllocateAll!(3008)(100, 3);
+
+    testAllocateAll!(1)(0, 128);
+    testAllocateAll!(1)(128 * 1, 128);
+    testAllocateAll!(128 * 20)(13 * 128, 128);
+}
+
+// Test totalAllocation
+unittest
+{
+    BitmappedBlock!(8, 8, NullAllocator) h1;
+    assert(h1.totalAllocation(1) >= 8);
+    assert(h1.totalAllocation(64) >= 64);
+    //writeln(h1.totalAllocation(8 * 64));
+    assert(h1.totalAllocation(8 * 64) >= 8 * 64);
+    assert(h1.totalAllocation(8 * 63) >= 8 * 63);
+    assert(h1.totalAllocation(8 * 64 + 1) >= 8 * 65);
+
+    BitmappedBlock!(64, 8, NullAllocator) h2;
+    assert(h2.totalAllocation(1) >= 64);
+    assert(h2.totalAllocation(64 * 64) >= 64 * 64);
+
+    BitmappedBlock!(4096, 4096, NullAllocator) h3;
+    assert(h3.totalAllocation(1) >= 4096);
+    assert(h3.totalAllocation(64 * 4096) >= 64 * 4096);
+    assert(h3.totalAllocation(64 * 4096 + 1) >= 65 * 4096);
+}
+
+// BitmappedBlockWithInternalPointers
+/**
+
+A $(D BitmappedBlock) with additional structure for supporting $(D
+resolveInternalPointer). To that end, $(D BitmappedBlockWithInternalPointers) adds a
+bitmap (one bit per block) that marks object starts. The bitmap itself has
+variable size and is allocated together with regular allocations.
+
+The time complexity of $(D resolveInternalPointer) is $(BIGOH k), where $(D k)
+is the size of the object within which the internal pointer is looked up.
+
+*/
+struct BitmappedBlockWithInternalPointers(
+    size_t theBlockSize, uint theAlignment = platformAlignment,
+    ParentAllocator = NullAllocator)
+{
+    import std.conv : text;
+    unittest
+    {
+        import std.experimental.allocator.mallocator : AlignedMallocator;
+        auto m = AlignedMallocator.instance.alignedAllocate(1024 * 64,
+            theAlignment);
+        testAllocator!(() => BitmappedBlockWithInternalPointers(m));
```

Comment:
The line `scope(exit) AlignedMallocator.instance.deallocate(m);` should be added here.